src/ptvsd/server/api.py

```diff
@@ -17,6 +17,44 @@
 from pydevd_file_utils import get_abs_path_real_path_and_base_from_file
 
 
+_ADAPTER_START_TIMEOUT = 10
+_dap_listener = None
+class _DAPMessagesListener(pydevd.IDAPMessagesListener):
+    def __init__(self):
+        self._adapter_port = None
+        self._adapter_port_event = threading.Event()
+
+    def before_send(self, msg):
+        pass
+
+    def after_receive(self, msg):
+        try:
+            if msg["command"] == "setDebuggerProperty":
+                self._adapter_port = msg["arguments"]["adapterPort"]
+                self._adapter_port_event.set()
+        except:
+            pass
+
+    @staticmethod
+    def get_adapter_port():
+        global _dap_listener
+        assert _dap_listener is not None
+        log.info("get_adapter_port()")
+        if not _dap_listener._adapter_port_event.wait(_ADAPTER_START_TIMEOUT):
+            log.error("get_adapter_port() timed out.")
+            raise TimeoutError("Failed to get adapter port.")
+        return _dap_listener._adapter_port
+
+    @staticmethod
+    def initialize():
+        global _dap_listener
+        if _dap_listener is not None:
+            return
+
+        _dap_listener = _DAPMessagesListener()
+        pydevd.add_dap_messages_listener(_dap_listener)
+
+
```

Comment:
This all can go inside `enable_attach()`, since it's its private implementation detail that's not used elsewhere.

Also, we really only support calling `enable_attach()` once inside the process already. It doesn't check for it at the moment, but it probably should do so, and raise an error for subsequent calls. That will then also ensure that only one instance of this object is created and registered by that first `enable_attach()` call, and so it doesn't need to do any checks itself.

To simplify it even further, consider using `Queue` (`from ptvsd.compat import queue`). You can create it as a local variable inside `enable_attach()`, and then have the listener `put()` the port number into it, and the caller `get()` it. Then this class would only need the pydevd callback methods, and nothing else.